### PR TITLE
Prototype clipboard api

### PR DESCRIPTION
**Pull request by EgorOrachyov**

### CODE REVIEW

File: shared/java/org/jetbrains/jwm/ClipboardEntry.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @NotNull @SneakyThrows
+    public static ClipboardEntry makeString(ClipboardFormat format, String text) {
+        return make(format, text.getBytes("UTF-16LE"));
+    }
+
+    @NotNull @SneakyThrows
+    public String getString() {
+        assert ClipboardFormat.isStringCompatible(_format);
```
Comment (tonsky) [inline shared/java/org/jetbrains/jwm/ClipboardEntry.java:38]: What if user register her own string-compatible format?

File: shared/java/org/jetbrains/jwm/Clipboard.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            return format;
+
+        if (!_nRegisterFormat(formatId))
+            throw new RuntimeException("Failed to register format: " + formatId);
+
+        return registerFormatInternal(formatId);
+    }
+
+    @ApiStatus.Internal @NotNull
+    public static ClipboardFormat registerFormatInternal(String formatId) {
```
Comment (tonsky) [inline shared/java/org/jetbrains/jwm/Clipboard.java:47]: Our convention is to use `_` to indicate internal methods/fields.

File: shared/java/org/jetbrains/jwm/Clipboard.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    public static ClipboardFormat registerFormat(String formatId) {
+        ClipboardFormat format = _formats.get(formatId);
+
+        if (format != null)
+            return format;
+
+        if (!_nRegisterFormat(formatId))
+            throw new RuntimeException("Failed to register format: " + formatId);
+
+        return registerFormatInternal(formatId);
```
Comment (tonsky) [inline shared/java/org/jetbrains/jwm/Clipboard.java:43]: Need some lock between checking if format exists and actually registering new format

**Review by tonsky — CHANGES_REQUESTED** [review]
Looks great, some minor details need polishing (see comments)

File: shared/java/org/jetbrains/jwm/Clipboard.java (hunk 3 above)
Comment (EgorOrachyov) [reply shared/java/org/jetbrains/jwm/Clipboard.java:43]: I thought, our api calls are allowed only on UI thread?

**Pull request merged**